src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -181,17 +181,36 @@ void ICollection.CopyTo(Array array, int index)
         }
 
         // Adds item to the tail of the queue.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Enqueue(T item)
         {
-            if (_size == _array.Length)
+            int size = _size;
+            T[] array = _array;
+
+            if ((uint)size < (uint)array.Length)
             {
-                int newcapacity = (int)((long)_array.Length * (long)GrowFactor / 100);
-                if (newcapacity < _array.Length + MinimumGrow)
-                {
-                    newcapacity = _array.Length + MinimumGrow;
-                }
-                SetCapacity(newcapacity);
+                _array[_tail] = item;
+                MoveNext(ref _tail);
+                _size++;
+                _version++;
+            }
+            else
+            {
+                EnqueueWithResize(item);
             }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnqueueWithResize(T item)
+        {
+            int newCapacity = _array.Length * 2;
+
+            if (newCapacity < _array.Length + MinimumGrow)
+            {
+                newCapacity = _array.Length + MinimumGrow;
```

Comment:
Unfortunately the JIT fetches `_array.Length` on each use. So cache it, like:
```c#
[MethodImpl(MethodImplOptions.NoInlining)]
private void EnqueueWithResize(T item)
{
    int length = _array.Length;
    int newCapacity = length * 2;
    int lengthWithGrow = length + MinimumGrow;

    if (newCapacity < lengthWithGrow)
    {
        newCapacity = lengthWithGrow;
    }

    SetCapacity(newCapacity);
    // ...
```

Reply:
Or use `Math.Min`:
```c#
int length = _array.Length;
int newCapacity = Math.Min(length * 2, length + MinimumGrow);

SetCapacity(newCapacity);
```

Reply:
PS: I know it's the cold path, but reduced asm-size is also benefitial :wink:

Reply:
> Unfortunately the JIT fetches _array.Length on each use. 

For the benefit of others: the JIT must do this since it cannot predict access from other threads...

Reply:
I didn't actually know that, thanks!

Reply:
the big deal 😄

Reply:
> I didn't actually know that

The nice part of PRs. You improve the product and may learn something 😄

Reply:
Hopefully my statement is correct then. @jkotas? I guess it wants to avoid a situation where the body runs indefinitely repeatedly checking the field for some expected external change that actually already happened.

Reply:
Nothing to do with multiple threads. It is missed optimization. There is nothing fundamental preventing the JIT from doing the transformation with caching the length in locals that you have done manually.

`lengthWithGrow` is not strictly needed with current JIT. The JIT is able to do the transformation for you.

> Or use `Math.Min`

You probably meant `Math.Max`. That is the most sensible and readable way to write this. It should generate about as good code as the manually inlined version.

Reply:
> `lengthWithGrow` is not strictly needed with current JIT.

Hm, I saw the JIT emitting two `lea`s for `length + MinimumGrow`, one in the `if` and one in the assignment. That's why I did it manually.

> You probably meant `Math.Max`

Oh, of couse.

Reply:
Made some updates here based on feedback